crates/flux-driver/src/collector/detached_specs.rs

```diff
@@ -0,0 +1,420 @@
+use std::collections::{HashMap, hash_map::Entry};
+
+use flux_common::span_bug;
+use flux_middle::fhir::Trusted;
+use flux_syntax::surface::{self, FnSpec, Item, Span};
+use rustc_errors::ErrorGuaranteed;
+use rustc_hir::{
+    OwnerId,
+    def::{DefKind, Res},
+    def_id::{CRATE_DEF_ID, LocalDefId},
+};
+use rustc_middle::ty::{AssocItem, AssocKind, Ty, TyCtxt};
+use rustc_span::{Symbol, def_id::DefId};
+
+use crate::collector::{FluxAttrs, SpecCollector, errors, surface::Ident};
+type Result<T = ()> = std::result::Result<T, ErrorGuaranteed>;
+
+#[derive(PartialEq, Eq, Debug, Hash)]
+struct ImplKey(Symbol);
+
+impl ImplKey {
+    fn new(ty: Ty) -> Self {
+        ImplKey(Symbol::intern(&format!("{ty:?}")))
+    }
+}
+
+type TraitImplKey = (ImplKey, Symbol);
+
+struct DetachedItems {
+    items: HashMap<Ident, (surface::Item, Option<DefId>)>,
+    inherent_impls: HashMap<Ident, (surface::DetachedInherentImpl, Option<DefId>)>,
+    trait_impls: HashMap<TraitImplKey, (surface::DetachedTraitImpl, Option<LocalDefId>, Span)>,
+}
+
+impl DetachedItems {
+    fn new(detached_specs: surface::DetachedSpecs, collector: &mut SpecCollector) -> Result<Self> {
+        let mut items = HashMap::default();
+        let mut inherent_impls: HashMap<Ident, (surface::DetachedInherentImpl, Option<DefId>)> =
+            HashMap::default();
+        let mut trait_impls = HashMap::default();
+        for item in detached_specs.items {
+            match item.kind {
+                surface::ItemKind::InherentImpl(detached_impl) => {
+                    if let Some(existing) = inherent_impls.get_mut(&item.ident) {
+                        existing.0.extend(detached_impl);
+                    } else {
+                        inherent_impls.insert(item.ident, (detached_impl, None));
+                    }
+                }
+                surface::ItemKind::TraitImpl(detached_impl) => {
+                    let key = (ImplKey(item.ident.name), detached_impl.trait_.name);
+                    let span = item.ident.span;
+                    if let std::collections::hash_map::Entry::Vacant(e) = trait_impls.entry(key) {
+                        e.insert((detached_impl, None, span));
+                    } else {
+                        return Err(collector.errors.emit(errors::AttrMapErr {
+                            span,
+                            message: format!("multiple impls for `{}`", item.ident.name),
+                        }));
```

Comment:
You shouldn't reuse `AttrMapErr` which is intended for errors when parsing attributes of the form `#[name(key1=val1,...)]`. Also, if you use this, the wording will be quite confusing because it says `invalid attribute: multiple impls for ident`, an talking about the detached spec as an attribute doesn't make much sense. This applies to all the places in this module where you use `AttrMapErr`

Reply:
indeed, have split into separate errors.